pubsub/google/cloud/pubsub_v1/publisher/client.py

```diff
@@ -0,0 +1,162 @@
+# Copyright 2017, Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+import copy
+import pkg_resources
+
+import six
+
+from google.cloud.gapic.pubsub.v1 import publisher_client
+
+from google.cloud.pubsub_v1 import _gapic
+from google.cloud.pubsub_v1 import types
+from google.cloud.pubsub_v1.publisher.batch import base
+from google.cloud.pubsub_v1.publisher.batch import thread
+
+
+__VERSION__ = pkg_resources.get_distribution('google-cloud-pubsub').version
+
+
+@_gapic.add_methods(publisher_client.PublisherClient, blacklist=('publish',))
+class Client(object):
+    """A publisher client for Google Cloud Pub/Sub.
+
+    This creates an object that is capable of publishing messages.
+    Generally, you can instantiate this client with no arguments, and you
+    get sensible defaults.
+
+    Args:
+        batch_settings (~.pubsub_v1.types.BatchSettings): The settings
+            for batch publishing.
+        batch_class (class): A class that describes how to handle
+            batches. You may subclass the
+            :class:`.pubsub_v1.publisher.batch.base.BaseBatch` class in
+            order to define your own batcher. This is primarily provided to
+            allow use of different concurrency models; the default
+            is based on :class:`threading.Thread`.
+        kwargs (dict): Any additional arguments provided are sent as keyword
+            arguments to the underlying
+            :class:`~.gapic.pubsub.v1.publisher_client.PublisherClient`.
+            Generally, you should not need to set additional keyword arguments.
+    """
+    def __init__(self, batch_settings=(), batch_class=thread.Batch, **kwargs):
+        # Add the metrics headers, and instantiate the underlying GAPIC
+        # client.
+        kwargs['lib_name'] = 'gccl'
+        kwargs['lib_version'] = __VERSION__
+        self.api = publisher_client.PublisherClient(**kwargs)
+        self.batch_settings = types.BatchSettings(*batch_settings)
+
+        # The batches on the publisher client are responsible for holding
+        # messages. One batch exists for each topic.
+        self._batch_class = batch_class
+        self._batches = {}
+
+        # Instantiate the "rejection batch", which is used for single-op
+        # acceptance checks if no batch is present.
+        self._rejection = base.RejectionBatch()
+
+    def batch(self, topic, message, create=True, autocommit=True):
```

Comment:
This function should be executed in a lock to prevent a race condition where a batch can be created and lost when another batch is created on the same topic.